core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala

```diff
@@ -277,12 +282,52 @@ private[spark] class CoarseGrainedExecutorBackend(
       if (executor != null) {
         executor.decommission()
       }
-      logInfo("Done decommissioning self.")
+      // Shutdown the executor once all tasks are gone & any configured migrations completed.
+      // Detecting migrations completion doesn't need to be perfect and we want to minimize the
+      // overhead for executors that are not in decommissioning state as overall that will be
+      // more of the executors. For example, this will not catch a block which is already in
+      // the process of being put from a remote executor before migration starts. This trade-off
+      // is viewed as acceptable to minimize introduction of any new locking structures in critical
+      // code paths.
+
+      val shutdownThread = new Thread("wait-for-blocks-to-migrate") {
+        var lastTaskRunningTime = System.nanoTime()
+        val sleep_time = 1000 // 1s
+
+        while (true) {
```

Comment:
We need to execute this while loop within the `Thread#run` and also start the thread. 

Currently the while runs as part of the construction of the new thread instance by the caller's thread and not as a separate new thread.

Reply:
NICE CATCH. OMG!!.

@holdenk, I think we should add tests for this :-)

Reply:
Yeah good catch. This will block on the decommissioning message and we want to return that. Let me think of how we can check that we’re not blocking in the decommission message.

Reply:
hmmm checking that this is non blocking is going to be a pain, and from what we've seen it works out ok even if it does block. I'll switch it to non-blocking, but if it works regardless of blocking or non-blocking do we need a test for this?

Reply:
I prefer to have a test (either a new one dedicated for this case or better to modify an existing one to check the executor really shutdowned).

Reply:
So the executor would shut down (we already test for that), it's just the RPC call would have been blocking rather than non-blocking.

Reply:
I see, you probably implying this test and this line:

https://github.com/apache/spark/blob/63846f1029c412376e5b414f8546f5d28453fa68/core/src/test/scala/org/apache/spark/storage/BlockManagerDecommissionIntegrationSuite.scala#L217

I had no time to run the test yet or check the details.

Reply:
Yeah. I'll double check by disabling the shutdown and seeing if the test fails in CI while I'm doing my other adventures today. (Then I'll either re-enable the shutdown code or work on getting it to fail first then re-enable it).

Reply:
I am just wondering there is no timeout for this. So if somebody in the future make a mistake this will hang as long as the Jenkins admin comes. Should not we need to use `tryAcquire`?

Reply:
Switched it to tryAcquire 👍